packages/dds/tree/src/feature-libraries/sequence-change-family/compose.ts

```diff
@@ -0,0 +1,205 @@
+        nextNewMark = undefined;
+        nextBaseMark = undefined;
+        if (baseMark === undefined) {
+            factory.push(clone(newMark));
+        } else if (isAttachGroup(newMark) || isReattach(newMark)) {
+            factory.pushContent(clone(newMark));
+            nextBaseMark = baseMark;
+        } else if (isDetachMark(baseMark)) {
+            // TODO: match base detaches to tombs and reattach in the newMarkList
+            factory.pushContent(baseMark);
+            nextNewMark = newMark;
+        } else {
+            const newMarkLength = getMarkLength(newMark);
+            const baseMarkLength = getMarkLength(baseMark);
+            if (newMarkLength < baseMarkLength) {
+                [baseMark, nextBaseMark] = splitMark(baseMark, newMarkLength);
+            } else if (newMarkLength > baseMarkLength) {
+                [newMark, nextNewMark] = splitMark(newMark, baseMarkLength);
+            }
+            // Passed this point, we are guaranteed that `newMark` and `baseMark` have the same length
+            if (typeof baseMark === "number") {
+                // TODO: insert new tombs and reattaches without replacing the offset
+                factory.push(newMark);
+            } else {
+                const composedMark = composeMarks(newMark, baseMark);
+                factory.push(...composedMark);
+            }
+        }
+        if (nextBaseMark === undefined) {
+            iBase += 1;
+            nextBaseMark = baseMarkList[iBase];
+        }
+        if (nextNewMark === undefined) {
+            iNew += 1;
+            nextNewMark = newMarkList[iNew];
+        }
+    }
+    if (nextBaseMark !== undefined) {
+        factory.push(nextBaseMark);
+    }
+    factory.push(...baseMarkList.slice(iBase + 1));
+    return factory.list;
+}
+
+function composeMarks(
+    newMark: T.SizedMark,
+    baseMark: T.ObjectMark | T.AttachGroup,
+): T.Mark[] {
+    if (typeof newMark === "number") {
+        return [baseMark];
+    }
+    const newType = newMark.type;
+    if (isAttachGroup(baseMark)) {
+        switch (newType) {
+            case "Modify": {
+                const attach = baseMark[0];
+                if (attach.type === "Insert") {
+                    return [[{
+                        ...newMark,
+                        type: "MInsert",
+                        id: attach.id,
+                        content: attach.content[0],
+                    }]];
+                } else if (attach.type === "MInsert") {
+                    updateModifyLike(newMark, attach);
+                    return [[attach]];
+                }
+                fail("Not implemented");
+            }
+            case "Delete": {
+                // The insertion of the previous change is subsequently deleted.
+                // TODO: preserve the insertion as muted
+                return [];
+            }
+            default: fail("Not implemented");
+        }
+    }
+    const baseType = baseMark.type;
```

Comment:
It would be nice if it were easier to see that `isAttachGroup(baseMark)` must be false when we reach this line. Maybe create a `composeWithAttach` helper.